src/smt/optimization_solver.h

```diff
@@ -122,32 +163,49 @@ class OptimizationSolver
    * Constructor
    * @param parent the smt_solver that the user added their assertions to
    **/
-  OptimizationSolver(SmtEngine* parent);
-  ~OptimizationSolver();
+  OptimizationSolver(SmtEngine* parent)
+      : d_parent(parent), d_objectives(), d_results()
+  {
+  }
+  ~OptimizationSolver() = default;
+
+  /** Runs the optimization loop for the pushed objective **/
+  OptimizationResult::ResultType checkOpt();
 
-  /** Runs the optimization loop for the activated objective **/
-  OptResult checkOpt();
   /**
-   * Activates an objective: will be optimized for
-   * @param obj the Node representing the expression that will be optimized for
+   * Pushes an objective: will be optimized for
+   * @param target the Node representing the expression that will be optimized
+   *for
    * @param type specifies whether it's maximize or minimize
    * @param bvSigned specifies whether we should use signed/unsigned
    *   comparison for BitVectors (only effective for BitVectors)
    *   and its default is false
    **/
-  void activateObj(const Node& obj,
-                   const ObjectiveType type,
-                   bool bvSigned = false);
-  /** Gets the value of the optimized objective after checkopt is called **/
-  Node objectiveGetValue();
+  void pushObjective(const Node target,
```

Comment:
can't this still be a reference?

Reply:
Node is already a reference to NodeValue. Using a reference of Node which is already a reference is redundant.

Reply:
No, it's not redudant since Node is essentially a shared_ptr to NodeValue and you don't want to increment the reference count each time `pushObjective` is called. Instead of using Node use either a `const Node&` or a `const TNode`.

Reply:
The parameter Node will be deallocated by RAII when the function is returned.

Reply:
and storing something in a vector needs strong reference.

Reply:
> The parameter Node will be deallocated by RAII when the function is returned.

That is correct but orthogonal to what @yoni206 and @mpreiner's point. The point is that creating/destroying the `Node` instance for the parameter is not necessary in the first place. The creation (resp. destruction) of nodes is not free: Each time we do that, we have to increment (resp. decrement) the reference count. Even if you end up storing the parameter in a vector, you still avoid unnecessary work by using a reference (see [this example](https://godbolt.org/z/qMojeP787)).

Reply:
@4tXJ7f Right! I'm porting all of them to TNode now, and the reason is performance-wise you said. Wait a sec for re-compilation on my side. 
To add to your point: The excessive performance of increasing the rc comes from the rc mutex I believe.

Reply:
@4tXJ7f oh, Node is not using shared_ptr internally and it's currently NOT thread-safe. 
Increasing the rc is simply increasing an integer in the current implementation.

Reply:
Ported to TNode for all function parameters (including constructors). Now this issue should be resolved.

Reply:
> @4tXJ7f oh, Node is not using shared_ptr internally and it's currently NOT thread-safe.
That's why @mpreiner wrote "essentially a shared_ptr". Does not change the fact that we do reference counting (albeit "manually" and not via a shared_ptr), which adds overhead.

Reply:
@aniemetz Logical! 
In the current implementation I suspect the only "overhead" in Node's rc is the addition of a few inlined integer instructions... it's minimal and definitely not the bottleneck. 
Whereas std::share_ptr's overhead mainly comes from its mutex.